src/category_theory/category/PartialFun.lean

```diff
@@ -0,0 +1,147 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import category_theory.category.Pointed
+import data.pfun
+
+/-!
+# The category of types with partial functions
+
+This defines `PartialFun`, the category of types equipped with partial functions.
+
+This category is classically equivalent to the category of pointed types. The reason it doesn't hold
+classically stems from the difference between `part` and `option`. Both can model partial functions,
+but the latter forces a decidable domain.
+
+## References
+
+* [nLab, *The category of sets and partial functions]
+[http://nlab-pages.s3.us-east-2.amazonaws.com/nlab/show/partial+function]
+-/
+
+open category_theory option
+
+universes u
+variables {α β : Type*}
+
+/-- The category of types equipped with partial functions. -/
+def PartialFun : Type* := Type*
+
+namespace PartialFun
+
+instance : has_coe_to_sort PartialFun Type* := ⟨id⟩
+
+/-- Turns a type into a `PartialFun`. -/
+@[nolint has_inhabited_instance] def of : Type* → PartialFun := id
+
+instance : inhabited PartialFun := ⟨Type*⟩
+
+instance large_category : large_category.{u} PartialFun :=
+{ hom := pfun,
+  id := pfun.id,
+  comp := λ X Y Z f g, g.comp f,
+  id_comp' := @pfun.comp_id,
+  comp_id' := @pfun.id_comp,
+  assoc' := λ W X Y Z _ _ _, (pfun.comp_assoc _ _ _).symm }
+
+end PartialFun
+
+/-- The forgetful functor from `Type` to `PartialFun` which forgets that the maps are total. -/
+def Type_to_PartialFun : Type ⥤ PartialFun :=
+{ obj := id,
+  map := @pfun.lift,
+  map_comp' := λ _ _ _ _ _, pfun.coe_comp _ _ }
+
+instance : faithful Type_to_PartialFun := ⟨λ X Y, pfun.coe_injective⟩
+
+/-- The functor which deletes the point of a pointed type. In return, this makes the maps partial.
+This the computable part of the equivalence `PartialFun_equiv_Pointed`. -/
+def Pointed_to_PartialFun : Pointed.{u} ⥤ PartialFun :=
+{ obj := λ X, {x : X // x ≠ X.point},
+  map := λ X Y f, pfun.to_subtype _ f.to_fun ∘ subtype.val,
+  map_id' := λ X, pfun.ext $ λ a b,
+    pfun.mem_to_subtype_iff.trans (subtype.coe_inj.trans part.mem_some_iff.symm),
+  map_comp' := λ X Y Z f g, pfun.ext $ λ a c, begin
+    refine (pfun.mem_to_subtype_iff.trans _).trans part.mem_bind_iff.symm,
+    simp_rw [pfun.mem_to_subtype_iff, subtype.exists],
+    refine ⟨λ h, ⟨f.to_fun a, λ ha, c.2 $ h.trans
+      ((congr_arg g.to_fun ha : g.to_fun _ = _).trans g.map_point), rfl, h⟩, _⟩,
+    rintro ⟨b, _, (rfl : b = _), h⟩,
+    exact h,
+  end }
+
+/-- The functor which maps undefined values to a new point. This makes the maps total and creates
+pointed types. This the noncomputable part of the equivalence `PartialFun_equiv_Pointed`. It can't
+be computable because `= option.none` is decidable while the domain of a general `part` isn't. -/
```

Comment:
Why is it relevant here that `= option.none` is decidable?

Reply:
It's relevant because you're turning a partial function `X →. Y` into a function `option X → option Y` by sending to `none` the undefined stuff. But being defined is (generally) undecidable while being sent to `none` is decidable. So it can't be constructive.

Reply:
Alright, can you add this to the explanation here?